solc/CommandLineInterface.cpp

```diff
@@ -845,52 +845,56 @@ void CommandLineInterface::handleCombinedJSON()
 	output[g_strVersion] = frontend::VersionString;
 	vector<string> contracts = m_compiler->contractNames();
 
+	// NOTE: The state checks here are more strict that in Standard JSON. There we allow
+	// requesting certain outputs even if compilation fails as long as analysis went ok.
+	bool compilationSuccess = m_compiler->state() >= CompilerStack::State::CompilationSuccessful;
+
 	if (!contracts.empty())
 		output[g_strContracts] = Json::Value(Json::objectValue);
 	for (string const& contractName: contracts)
 	{
 		Json::Value& contractData = output[g_strContracts][contractName] = Json::objectValue;
-		if (m_options.compiler.combinedJsonRequests->abi)
+		if (m_options.compiler.combinedJsonRequests->abi && compilationSuccess)
 			contractData[g_strAbi] = m_compiler->contractABI(contractName);
-		if (m_options.compiler.combinedJsonRequests->metadata)
+		if (m_options.compiler.combinedJsonRequests->metadata && compilationSuccess)
 			contractData["metadata"] = m_compiler->metadata(contractName);
```

Comment:
I decided to fix the ICE from https://github.com/ethereum/solidity/pull/14323#discussion_r1228292931 after all, since I had to update the existing conditions anyway and the fix is just to add the check to outputs that did not have it.